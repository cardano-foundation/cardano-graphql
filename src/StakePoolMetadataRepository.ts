import * as path from 'path'
import * as fs from 'fs-extra'
import * as simpleGit from 'simple-git/promise'
import { StakePool } from './graphql_types'
const lowDb = require('lowdb')
const Memory = require('lowdb/adapters/Memory')

export type Config = {
  cloneOptions?: { [key: string]: null | string }
  localPath: string,
  remoteUri: string
}

export function StakePoolMetadataRepository ({ cloneOptions, localPath, remoteUri }: Config) {
  let git: ReturnType<typeof simpleGit>
  let db: ReturnType<typeof lowDb>
  async function stakePoolData (file: string) {
    const entry = await fs.readJson(path.join(localPath, file))
    return {
      ...entry,
      ...{ id: path.parse(file).name }
    }
  }

  return {
    destroy () {
      return fs.remove(localPath)
    },
    get (id: string): Promise<Partial<StakePool>> {
      return db.get('stakePools').find({ id }).value()
    },
    async has (id: string) {
      return !!await this.get(id)
    },
    async init () {
      db = lowDb(new Memory())
      if (!await fs.pathExists(localPath)) {
        git = simpleGit()
        await git.clone(remoteUri, localPath, cloneOptions)
      } else {
        git = simpleGit(localPath)
        await this.pull()
      }
      const entries = (await fs.readdir(localPath))
        .filter(isMetadataFile)
        .map(stakePoolData)
      db.defaults({ stakePools: await Promise.all(entries) }).write()
    },
    async pull () {
      await git.cwd(localPath)
      return git.pull('origin', 'master')
    },
    size () {
      return db.get('stakePools').size().value()
    },
    async syncWithRemote () {
      const res = await this.pull()
      const { created, deleted, files } = res
      if (created.length > 0) {
        const stakePools = created.filter(isMetadataFile)
        for (const file of stakePools) {
          db.get('stakePools').push(await stakePoolData(file)).write()
        }
      } else if (deleted.length > 0) {
        const stakePools = deleted.filter(isMetadataFile)
        for (const file of stakePools) {
          db.get('stakePools').remove({ id: path.parse(file).name }).write()
        }
      } else if (files.length > 0) {
        const stakePools = files.filter(isMetadataFile)
        for (const file of stakePools) {
          db.get('stakePools').find({ id: path.parse(file).name })
            .assign(await stakePoolData(file)).write()
        }
      }
    }
  }
}

function isMetadataFile (file: string): boolean {
  return !/^\..*/.test(file) && path.extname(file).toLowerCase() === '.json'
}
