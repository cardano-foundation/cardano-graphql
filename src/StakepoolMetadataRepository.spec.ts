import * as path from 'path'
import * as fs from 'fs-extra'
import * as simpleGit from 'simple-git/promise'
import { StakePoolMetadataRepository } from './StakePoolMetadataRepository'

const git = simpleGit()
async function addStakePoolToRemoteRepo (id: string, remoteUri: string) {
  await fs.writeJson(`sp_${id}.json`, {
    description: `A stakepool with the ID of ${id}`,
    isCharity: true,
    profitMargin: 30,
    name: `Stake Pool ${id}`,
    ticker: `SP${id}`,
    url: `http://stake.pool/${id}`
  })
  await git.add('./*')
  await git.commit(`Add stake pool ${id}`)
  await git.addRemote('origin', remoteUri)
  return git.push()
}

describe('StakePoolMetadataRepository', () => {
  let metadataRepository: ReturnType<typeof StakePoolMetadataRepository>
  const LOCAL_REPO_PATH = path.join(__dirname, '__temp_metadata_repo__')
  const REMOTE_REPO_URI = 'http://localhost:4040/stake-pool-metadata.git'

  const localRepoExists = () =>  fs.pathExists(LOCAL_REPO_PATH)

  beforeEach(async () => {
    metadataRepository = StakePoolMetadataRepository({
      localPath: LOCAL_REPO_PATH,
      remoteUri: REMOTE_REPO_URI
    })
   await metadataRepository.destroy()
  })

  afterEach(async () => await metadataRepository.destroy())

  describe('init', () => {

    it('Ensures the directory exists and clones the remote repository if needed', async () => {
      expect(await localRepoExists()).toBe(false)
      await metadataRepository.init()
      expect(await localRepoExists()).toBe(true)
      expect(await fs.pathExists(path.join(LOCAL_REPO_PATH, 'stake-pool-metadata','.git')))
    })
    it('pulls any changes if the local repository already exists', async () => {
      expect(await localRepoExists()).toBe(false)
      await metadataRepository.init()
      expect(await localRepoExists()).toBe(true)
      await metadataRepository.init()
      await addStakePoolToRemoteRepo("4", REMOTE_REPO_URI)
      expect(await fs.readFile(path.join(LOCAL_REPO_PATH, 'stake-pool-metadata','sp_4.json')))
    })
  })

  // describe('get', () => {
  //
  //   beforeEach(() => {
  //     metadataRepository.init()
  //   })
  //
  //   it('returns the json for the matching stake pool ID', () => {
  //
  //   })
  //   it('throws a 404 if no match is found', () => {
  //
  //   })
  // })

})
