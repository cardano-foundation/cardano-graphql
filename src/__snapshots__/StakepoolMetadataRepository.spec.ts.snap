// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakePoolMetadataRepository get returns the Stake Pool metadata for the matching stake pool ID 1`] = `
Object {
  "description": "A stakepool with the ID of sp1",
  "id": "sp1",
  "isCharity": false,
  "name": "Stake Pool SP1",
  "profitMargin": 30,
  "ticker": "SP1",
  "url": "http://stake.pool/sp1",
}
`;
